pipeline/alignments/align.py

```diff
@@ -153,17 +192,105 @@ def write_priors(
     priors_output_path: str,
-    stack: ExitStack,
 ):
+    import eflomal
+
     logger.info("Calculating priors...")
-    src_input = stack.enter_context(open(corpus_src, "r", encoding="utf-8"))
-    trg_input = stack.enter_context(open(corpus_trg, "r", encoding="utf-8"))
-    fwd_f = stack.enter_context(open(fwd_path, "r", encoding="utf-8"))
-    rev_f = stack.enter_context(open(rev_path, "r", encoding="utf-8"))
-    priors_tuple = eflomal.calculate_priors(src_input, trg_input, fwd_f, rev_f)
-    logger.info(f"Writing priors to {priors_output_path}...")
-    priors_output = stack.enter_context(open(priors_output_path, "w", encoding="utf-8"))
-    eflomal.write_priors(priors_output, *priors_tuple)
+    with ExitStack() as stack:
+        src_input = stack.enter_context(open(corpus_src, "r", encoding="utf-8"))
+        trg_input = stack.enter_context(open(corpus_trg, "r", encoding="utf-8"))
+        fwd_f = stack.enter_context(open(fwd_path, "r", encoding="utf-8"))
+        rev_f = stack.enter_context(open(rev_path, "r", encoding="utf-8"))
+        priors_tuple = eflomal.calculate_priors(src_input, trg_input, fwd_f, rev_f)
+        logger.info(f"Writing priors to {priors_output_path}...")
+        priors_output = stack.enter_context(open(priors_output_path, "w", encoding="utf-8"))
+        eflomal.write_priors(priors_output, *priors_tuple)
+
+
+def remap(
+    src_path: str,
+    trg_path: str,
+    tok_src_path: str,
+    tok_trg_path: str,
+    aln_path: str,
+    output_aln_path: str,
+) -> None:
+    """
+    Remaps alignments that were calculated for Moses-tokenized corpus to whitespace-tokenized ones.
+    :param src_path: path to whitespace-tokenized sentences in source language
+    :param trg_path: path to whitespace-tokenized sentences in target language
+    :param tok_src_path: path to Moses-tokenized sentences in source language
+    :param tok_trg_path: path to Moses-tokenized sentences in target language
+    :param aln_path: path to the alignments calculated for Moses-tokenized corpus
+    :param output_aln_path: path to output alignments file remapped to whitespace-tokenized corpus
+    """
+    logger.info("Remapping alignments to whitespace tokenization")
+
+    with ExitStack() as stack:
+        pool = stack.enter_context(multiprocessing.Pool(processes=multiprocessing.cpu_count()))
+        # Buffering helps to minimize IO operations which speeds thing up significantly
+        output = stack.enter_context(open(output_aln_path, "w", buffering=500000))
+
+        lines = zip(
+            stack.enter_context(open(src_path)),
+            stack.enter_context(open(trg_path)),
+            stack.enter_context(open(tok_src_path)),
+            stack.enter_context(open(tok_trg_path)),
+            stack.enter_context(open(aln_path)),
+        )
+
+        # send lines to worker processes in chunks
+        for aln in tqdm(pool.imap(remap_line, lines, chunksize=10000), mininterval=10):
+            output.write(aln)
+
```

Comment:
Thought:

Something to measure process memory on tasks with many steps could be interesting.

Something like:

```
import psutil


def print_memory():
    processes = []

    # Collect process information
    for proc in psutil.process_iter(["pid", "name", "memory_info"]):
        try:
            pid = proc.info["pid"]
            name = proc.info["name"]
            memory_info = proc.info["memory_info"]
            memory_usage = memory_info.rss
            processes.append((name, pid, memory_usage))
        except Exception:
            print("Failed to get process information")

    # Sort the processes based on memory usage (descending)
    processes.sort(key=lambda x: x[2], reverse=True)

    # Calculate the maximum length of "name (pid)" for alignment
    display_names = [f"{name} ({pid})" for name, pid, _ in processes]
    max_length = max(len(display_name) for display_name in display_names)

    # Print the sorted process information with right-padding
    for proc_info, display_name in zip(processes, display_names):
        _, _, memory_usage = proc_info
        print(f"{display_name.ljust(max_length)} {memory_usage / (1024 * 1024):.2f} MB")


if __name__ == "__main__":
    print_memory()
```

```
plugin-container (6622)                              1081.06 MB
firefox (6609)                                       963.39 MB
plugin-container (6618)                              845.61 MB
plugin-container (6619)                              734.81 MB
iTerm2 (88484)                                       540.56 MB
Slack Helper (Renderer) (49150)                      515.27 MB
Code Helper (Renderer) (47144)                       450.91 MB
plugin-container (81278)                             438.48 MB
plugin-container (6621)                              434.69 MB
plugin-container (17813)                             336.47 MB
plugin-container (6620)                              333.00 MB
```

Reply:
Agreed, proper measuring would definitely be interesting but since the main issue is inside C++ code in eflomal, we can't do much about it anyway. We can run such a tool locally on a 10M dataset if you're interested.